### ISSUE TITLE
Invalid/Erratic Transactions get committed to the BlockChain

### ISSUE THREAD

**Issue by mcr222** [hyperledger-archives/fabric#983]
Currently invalid or wrong invoke transactions get committed the blockchain.
- Invalid transactions: tx referencing not deployed chaincode, calling chaincode without the correct credentials, calling a non-existent chaincode function...
- Erratic transactions: tx that can't be executed because of any random execution error or because it can't be performed with current values of world state variables (for example if money transactions are asked but account has insufficient balance).

In the current BlockChain design, all this "bad" transactions are still committed to the BC (although they obviously don't modify the world state variables).

This raises several problems because ultimately, the client won't know if his transaction has been executed. For example: a user does a transaction of 5€ but he only has 3€ in his account. How can this user know that the transaction was wrong if it was committed anyway and the user can check that it is registered in the BC? 

Here are two suggestions of what could be done (it should be thought thoroughly as it is a premature idea):
- When executing transactions (that have been previously ordered) each node does not commit those transactions that are invalid or return an error of execution. This would work if it is assumed that execution output is the same in all nodes (assumption that current BlockChain implementation assumes)
- Use Sieve consensus algorithm, which agrees on the output of transaction execution, if the transaction's execution output is an error or invalid discard transaction as it would be done if outputs do not match (I think this solution can be more easily implemented).

**Comment by bcbrock** [comment]
My first (rhetorical) question is: Why do you say that bad transactions don't modify the state variables? Where is this documented/specified? If a transaction does writes to the state and then exits with an error there is no specification of what happens or is supposed to happen. (Note, I'm griping about the lack of good specifications, not about your issue.)

I do disagree with one of your points, however. I would argue that any transaction that is accepted asynchronously (as they all are currently) _must_ be registered in the Blockchain. If the INVOKE interface has given the client a UUID, then the client must be able to unequivocally determine that the transaction UUID has been executed or will never be executed, even if the execution resulted in a failure. Simply omitting the UUID if it fails creates more problems than it solves in my view, since this prohibits the client from getting absolute "closure" on the state of its transactions. If HyperLedger included an interface that would block until the final disposition of the transaction (error or committed) was known then it would be OK to drop, I think.

**Comment by zmanian** [comment]
My understanding that if a validator node's chaincode returns an error on a transaction. The transaction will not being in the blockchain.

**Comment by mcr222** [comment]
@bcbrock I agree with what you "asked". It is true that if it is decided that wrong transactions need to be removed then it must be ensured that the transaction hasn't modified any state variable. It is a crucial observation.

I also agree that if the client doesn't know what happens with his/her transaction he/she might think that something went wrong (e.g. network down) instead of being sure that the transaction was wrong (e.g. balance insufficient). In any case, the client currently doesn't have any way of knowing if his/her transaction was wrong as the client will always see the transaction committed.

In my opinion transactions should be discarded before assigning them a total order through consensus (2nd of my suggested solutions).

**Comment by kletkeman** [comment]
@bcbrock @mcr222 I would be more comfortable if all incoming transactions were logged in the blockchain for audit purposes. Good or bad, anything that is attempted is a transaction and worthy of knowing about. I am of course presuming that defence against various attacks such as (D)DOS is handled outside the fabric.

**Comment by mcr222** [comment]
How would then any service on top of the blockchain know if their request (invoke transaction) was executed correctly? Many things can go wrong (reference a non-existent deploy tx, sending incorrect function parameters, parameters that the chaincode rejects...) and the transaction would still be registered as any other correct transaction.

**Comment by chetmurthy** [comment]
@bcbrock: you're making two requirements -- want to clarify, b/c
they're different, and one of them -might- or might not be good to
have.

(1) [clearly necessary] When a client submits a tran to a peer, it
should also be able to (by default YES, can't imagine NO) create a
local subscription in the peer, for thay tran-id.  Eventually (barring
certain problems, read below) that tran comes thru in the log via
consensus, and is either successfully applied, or aborted.  At that
point, the local peer delivers a notifcation of this, to that
subscribed client.

The creation of such a subscription by a client must be possible
independently of the submission of a tran, so that a client can
subscribe to the result of a tran it has yet to submit, at many
peers. This would be used to implement "strong reads".

This is uncontroversial.

The only snag in the above, is that the network might "lose track of"
a tran, and it never comes thru in the log.  This is a distributed
system -- such things happen.  We can look into all the edge cases
where it might happen, but I'm a little dour that we can find all of
them.

A non-snag: if the client crashes (or the peer crashes) that
subscription will be gone, and this is an expected thing -- if your
connection to DB2 crashes, you can't find out whether that tran on
which you just typed "COMMIT" actually committed.

(2) [somewhat controversial] If we keep all trans (both committed and
aborted) in the blockchain, it could be quite large.  Also, does this
mean that even trans that fail basic (local) consistency checks should
be stored ("everywhere") in the blockchain?

Could be significant storage.  That said, the -implementation- of this
is uncontroversial.

**Comment by bcbrock** [comment]
@mcr222 - I think the simple answer to one of your concerns is to include a status field in the transaction records in the Blockchain that indicates whether the transaction succeeded or failed, or more specifically how it succeeded or failed.

@chetmurthy - I agree that haviing success/failure status come back in an event stream seems like an appealing idea. This notification is ephemeral however - if you blink you can miss it. If this is the only notification for failed transactions it would require the client applications to probably be even more robust that the server, which seems upside-down to me.

I don't think the idea of failed transactions taking up too much space would really be an issue in a permissioned blockchain. Clients are not going to be simply "trying stuff to see what works".  The peers can probably also do more to filter out obviously bogus transactions, like those that refer to non-existent chain codes, before returning a transaction ID (i.e., before officially accepting the transaction).

Or, let's say there _is_ a rogue client. What is the the difference between the client sending 1,000,000 transactions that "work", are recorded in the chain but don't actually change the world state, vs. sending 1,000,000 that outright "fail"?

**Comment by chetmurthy** [comment]
@bcbrock SGTM  I just wanted to be sure the question of size of data was considered-and-dismissed.  Assumed it'd go this way, and I sure support.

**Comment by mcr222** [comment]
@bcbrock I agree that a status field would solve the issue of not knowing what happened with your transaction.

**Comment by srderson** [comment]
Related to PR #1719

**Comment by christo4ferris** [comment]
@srderson #1719 was merged, can we close this?
